### PR TITLE
Easier way to dnstap?

### PR DESCRIPTION
**Pull request by 2buisson**
### 1. What does this pull request do?

Does making dnstap easier to use in a plugin by putting the error handling outside of the plugin itself.

### 2. Which issues (if any) are related?

* https://github.com/coredns/coredns/issues/1448

### 3. Which documentation changes (if any) need to be made?

The Tapper interface in the context has changed.

### CODE REVIEW

File: plugin/dnstap/handler.go [hunk 1]
```diff
@@ -60,14 +62,19 @@ func tapMessageTo(w io.Writer, m *tap.Message) error {
...
+
 	h.IO.Dnstap(msg.Wrap(m))
-	return nil
+	return
 }
 
-// TapBuilder implements Tapper.
-func (h Dnstap) TapBuilder() msg.Builder {
-	return msg.Builder{Full: h.Pack}
+func (h Dnstap) Pack() bool {
```
Comment (miekg) [inline plugin/dnstap/handler.go]: nit: entire function can be put on a single line 
non nit: docs

File: plugin/dnstap/handler.go [hunk 2]
```diff
@@ -15,9 +15,11 @@ import (
 
 // Dnstap is the dnstap handler.
 type Dnstap struct {
-	Next plugin.Handler
-	IO   IORoutine
-	Pack bool
+	Next           plugin.Handler
+	IO             IORoutine
+	JoinRawMessage bool
```
Comment (miekg) [inline plugin/dnstap/handler.go]: JoinRawMessage should probably have some docs explaining what it does when you set it.

File: plugin/dnstap/msg/msg.go [hunk 3]
```diff
@@ -10,100 +9,89 @@ import (
...
-// Builder helps to build Data by being aware of the dnstap plugin configuration.
 type Builder struct {
-	Full bool
-	Data
+	Packed      []byte
+	SocketProto tap.SocketProtocol
+	SocketFam   tap.SocketFamily
+	Address     []byte
+	Port        uint32
+	TimeSec     uint64
```
Comment (miekg) [inline plugin/dnstap/msg/msg.go]: Can TimeSec be a `time.Time`?
or does everything here needs to be in a lower level format for easier packing?

Same question for Address (i.e. why not  an `net.IP`?)

And Port, why not just `int`?

File: plugin/dnstap/msg/msg.go [hunk 4]
```diff
@@ -10,100 +9,89 @@ import (
...
-// Data helps to build a dnstap Message.
-// It can be transformed into the actual Message using this package.
-type Data struct {
-	Packed      []byte
-	SocketProto tap.SocketProtocol
-	SocketFam   tap.SocketFamily
-	Address     []byte
-	Port        uint32
-	TimeSec     uint64
+func (b *Builder) Msg(m *dns.Msg) *Builder {
```
Comment (miekg) [inline plugin/dnstap/msg/msg.go]: Docs

File: plugin/dnstap/msg/msg.go [hunk 5]
```diff
@@ -10,100 +9,89 @@ import (
...
+		return b
+	}
+
+	b.Packed, b.err = m.Pack()
+	return b
 }
 
-// HostPort decodes into Data any string returned by dnsutil.ParseHostPortOrFile.
-func (d *Data) HostPort(addr string) error {
+func (b *Builder) HostPort(addr string) *Builder {
```
Comment (miekg) [inline plugin/dnstap/msg/msg.go]: Docs

**Review by miekg — CHANGES_REQUESTED** [review]
Welcome back! :)

Could you also document (in the README.md?) the easier way to use this?

File: plugin/dnstap/msg/msg.go (hunk 3 above)
Comment (2buisson) [reply plugin/dnstap/msg/msg.go]: Yes it is for the Dnstap struct to take the address of those fields, maybe they shouldn't be exported.

**Comment by miekg** [comment]
You can still use net.IP for instance.

**Comment by 2buisson** [comment]
Oops something is wrong do not merge!

**Comment by 2buisson** [comment]
Hopefully it's all good now :grin:

**Comment by miekg** [comment]
On holiday, trying to get a review done in the next few days.

**Review by miekg — APPROVED** [review]

**Pull request merged**